Don't support relationship indexes in apoc.export.cypher.all

Hi
I tried to export the database using the following query:

`CALL apoc.export.cypher.all("db.cypher",{format: "neo4j-shell", useOptimizations: {type: "UNWIND_BATCH",unwindBatchSize: 20}})`
After running, I encountered the following error:

> Failed to invoke procedure 'apoc.export.cypher.all' :  Caused by: java.lang.IllegalStateException: This is not a node index.

It seems that the export procedure only supports node indexes. So what about the relationship indexes?
### Versions

  - OS: Linux 20.04
  - Neo4j: 4.3.1
  - Neo4j-Apoc: 4.2.0.4

--- a/core/src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java
+++ b/core/src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java
@@ -246,8 +246,9 @@ public class MultiStatementCypherSubGraphExporter {
                         return null;  // delegate to the constraint creation
                     }
 
+                    boolean isNode = "NODE".equals(map.get("entityType"));
                     if ("FULLTEXT".equals(map.get("type"))) {
-                        if ("NODE".equals(map.get("entityType"))) {
+                        if (isNode) {
                             List<Label> labels = toLabels(tokenNames);
                             return this.cypherFormat.statementForNodeFullTextIndex(name, labels, props);
                         } else {
@@ -257,7 +258,11 @@ public class MultiStatementCypherSubGraphExporter {
                     }
                     // "normal" schema index
                     String tokenName = tokenNames.get(0);
-                    return this.cypherFormat.statementForIndex(tokenName, props);
+                    if (isNode) {
+                        return this.cypherFormat.statementForNodeIndex(tokenName, props);
+                    } else {
+                        return this.cypherFormat.statementForIndexRelationship(tokenName, props);
+                    }
 
                 })
                 .filter(StringUtils::isNotBlank)

--- a/core/src/main/java/apoc/export/cypher/formatter/AbstractCypherFormatter.java
+++ b/core/src/main/java/apoc/export/cypher/formatter/AbstractCypherFormatter.java
@@ -45,8 +45,13 @@ abstract class AbstractCypherFormatter implements CypherFormatter {
 	}
 
 	@Override
-	public String statementForIndex(String label, Iterable<String> keys) {
+	public String statementForNodeIndex(String label, Iterable<String> keys) {
 		return "CREATE INDEX ON :" + Util.quote(label) + "(" + CypherFormatterUtils.quote(keys) + ");";
+	}
+	
+	@Override
+	public String statementForIndexRelationship(String key, Iterable<String> keys) {
+		return String.format("CREATE INDEX FOR ()-[rel:%s]-() ON (%s);", Util.quote(key), getPropertiesQuoted(keys, "rel."));
 	}
 
 	@Override
@@ -80,11 +85,16 @@ abstract class AbstractCypherFormatter implements CypherFormatter {
 	@Override
 	public String statementForConstraint(String label, Iterable<String> keys) {
 
-		String keysString = StreamSupport.stream(keys.spliterator(), false)
-				.map(key -> "node." + CypherFormatterUtils.quote(key))
-				.collect(Collectors.joining(", "));
+		String keysString = getPropertiesQuoted(keys, "node.");
 
 		return  String.format(STATEMENT_CONSTRAINTS, Util.quote(label), keysString, Iterables.count(keys) > 1 ? "IS NODE KEY" : "IS UNIQUE");
+	}
+
+	private String getPropertiesQuoted(Iterable<String> keys, String prefix) {
+		String keysString = StreamSupport.stream(keys.spliterator(), false)
+				.map(key -> prefix + CypherFormatterUtils.quote(key))
+				.collect(Collectors.joining(", "));
+		return keysString;
 	}
 
 	protected String mergeStatementForNode(CypherFormat cypherFormat, Node node, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties, Set<String> indexNames) {

--- a/core/src/main/java/apoc/export/cypher/formatter/AbstractCypherFormatter.java
+++ b/core/src/main/java/apoc/export/cypher/formatter/AbstractCypherFormatter.java
@@ -50,8 +50,8 @@ abstract class AbstractCypherFormatter implements CypherFormatter {
 	}
 	
 	@Override
-	public String statementForIndexRelationship(String key, Iterable<String> keys) {
-		return String.format("CREATE INDEX FOR ()-[rel:%s]-() ON (%s);", Util.quote(key), getPropertiesQuoted(keys, "rel."));
+	public String statementForIndexRelationship(String type, Iterable<String> keys) {
+		return String.format("CREATE INDEX FOR ()-[rel:%s]-() ON (%s);", Util.quote(type), getPropertiesQuoted(keys, "rel."));
 	}
 
 	@Override

--- a/core/src/main/java/apoc/export/cypher/formatter/AddStructureCypherFormatter.java
+++ b/core/src/main/java/apoc/export/cypher/formatter/AddStructureCypherFormatter.java
@@ -38,7 +38,7 @@ public class AddStructureCypherFormatter extends AbstractCypherFormatter impleme
 	}
 	
 	@Override
-	public String statementForIndexRelationship(String label, Iterable<String> key) {
+	public String statementForIndexRelationship(String type, Iterable<String> key) {
 		return "";
 	}
 

--- a/core/src/main/java/apoc/export/cypher/formatter/AddStructureCypherFormatter.java
+++ b/core/src/main/java/apoc/export/cypher/formatter/AddStructureCypherFormatter.java
@@ -33,7 +33,12 @@ public class AddStructureCypherFormatter extends AbstractCypherFormatter impleme
 	}
 
 	@Override
-	public String statementForIndex(String label, Iterable<String> key) {
+	public String statementForNodeIndex(String label, Iterable<String> key) {
+		return "";
+	}
+	
+	@Override
+	public String statementForIndexRelationship(String label, Iterable<String> key) {
 		return "";
 	}
 

--- a/core/src/main/java/apoc/export/cypher/formatter/CypherFormatter.java
+++ b/core/src/main/java/apoc/export/cypher/formatter/CypherFormatter.java
@@ -19,7 +19,9 @@ public interface CypherFormatter {
 
 	String statementForRelationship(Relationship relationship, Map<String, Set<String>> uniqueConstraints, Set<String> indexedProperties);
 
-	String statementForIndex(String label, Iterable<String> keys);
+	String statementForNodeIndex(String label, Iterable<String> keys);
+
+	String statementForIndexRelationship(String type, Iterable<String> keys);
 
 	String statementForNodeFullTextIndex(String name, Iterable<Label> labels, Iterable<String> keys);
 

--- a/core/src/main/java/apoc/export/cypher/formatter/UpdateStructureCypherFormatter.java
+++ b/core/src/main/java/apoc/export/cypher/formatter/UpdateStructureCypherFormatter.java
@@ -33,7 +33,12 @@ public class UpdateStructureCypherFormatter extends AbstractCypherFormatter impl
 	}
 
 	@Override
-	public String statementForIndex(String label, Iterable<String> key) {
+	public String statementForNodeIndex(String label, Iterable<String> key) {
+		return "";
+	}
+
+	@Override
+	public String statementForIndexRelationship(String type, Iterable<String> key) {
 		return "";
 	}
 

--- a/core/src/test/java/apoc/export/cypher/ExportCypherTest.java
+++ b/core/src/test/java/apoc/export/cypher/ExportCypherTest.java
@@ -729,6 +729,20 @@ public class ExportCypherTest {
                     assertEquals(expected, unwind);
                 });
     }
+    
+    @Test
+    public void shouldSaveCorrectlyRelIndexesOptimized() throws FileNotFoundException {
+        String fileName = "relIndex.cypher";
+        db.executeTransactionally("CREATE INDEX rel_index_name FOR ()-[r:KNOWS]-() ON (r.since, r.foo)");
+        TestUtil.testCall(db, "CALL apoc.export.cypher.all($file, $exportConfig)",
+                map("file", fileName, "exportConfig", map("separateFiles", true, "format", "neo4j-shell")),
+                (r) -> assertResultsOptimized(fileName, r));
+        assertEquals(EXPECTED_NODES_OPTIMIZED, readFile("relIndex.nodes.cypher"));
+        assertEquals(EXPECTED_RELATIONSHIPS_OPTIMIZED, readFile("relIndex.relationships.cypher"));
+        assertEquals(EXPECTED_CLEAN_UP, readFile("relIndex.cleanup.cypher"));
+        assertEquals(EXPECTED_SCHEMA_WITH_RELS_OPTIMIZED, readFile("relIndex.schema.cypher"));
+        db.executeTransactionally("DROP INDEX rel_index_name");
+    }
 
     private void assertResultsOptimized(String fileName, Map<String, Object> r) {
         assertEquals(7L, r.get("nodes"));
@@ -909,6 +923,15 @@ public class ExportCypherTest {
                 "BEGIN%n" +
                 "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
                 "COMMIT%n");
+        
+        static final String EXPECTED_SCHEMA_WITH_RELS_OPTIMIZED = String.format("BEGIN%n" +
+                "CREATE INDEX ON :Bar(first_name,last_name);%n" +
+                "CREATE INDEX ON :Foo(name);%n" +
+                "CREATE INDEX FOR ()-[rel:KNOWS]-() ON (rel.since, rel.foo);%n" +
+                "CREATE CONSTRAINT ON (node:Bar) ASSERT (node.name) IS UNIQUE;%n" +
+                "CREATE CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT (node.`UNIQUE IMPORT ID`) IS UNIQUE;%n" +
+                "COMMIT%n" +
+                "SCHEMA AWAIT%n");
 
         static final String EXPECTED_SCHEMA_OPTIMIZED = String.format("BEGIN%n" +
                 "CREATE INDEX ON :Bar(first_name,last_name);%n" +

